### ISSUE TITLE
Feature Request: Support for private marketplace/gallery

### ISSUE THREAD

**Issue by levonmamikonyan** [microsoft/vscode#21839]
We have created several VSIX extension that have no use to anybody else except our company. We would like to host our own private extension gallery and have an ability to specify alternative extension gallery paths (like "Additional Extension Galleries" in Visual Studio 2015).

- VSCode Version: 1.10.1
- OS Version: Windows 10

**Comment by seanmcbreen** [comment]
Hi,

Currently we don't have a way to do this but it's not the first time we have heard this request so adding the correct label.  We don't have this in our current backlog but this will help us keep track of it.

Thanks
Sean

**Comment by hilleer** [comment]
This is a feature that we would welcome very, very much as well, as it's kind of time consuming and annoying to have to build and forward the vsix file every time you make changes :-)

**Comment by hilleer** [comment]
@seanmcbreen are there any updates on this request? :-)

**Comment by noxarivis** [comment]
@hilleer
nope, I asked recently

**Comment by hilleer** [comment]
Too bad....

**Comment by hilleer** [comment]
Anyone got updates on this?

**Comment by ibigpapa** [comment]
So viewing the code don't think i'll have time anytime soon to submit pull request.

https://github.com/Microsoft/vscode/blob/67cd9258fd0ed429ddc626d488d48bd65c2f2283/build/lib/extensions.js#L48

https://github.com/Microsoft/vscode/blob/67cd9258fd0ed429ddc626d488d48bd65c2f2283/build/lib/extensions.js#L67

https://github.com/Microsoft/vscode/blob/67cd9258fd0ed429ddc626d488d48bd65c2f2283/build/lib/extensions.js#L75

These appear to be the main lines used for getting extensions from the market place.  I guess you could replace the url with a call to the config to pull the url and insert a local cached proxy so it would pull from the local proxy if package is there if not pull from marketplace.  Then just provide the functions for the query and the upload that mimic the calls from vscode.

Cons to this would be https issues.

ideally you would modify the Extensions: Market Place UI to provide a filter to a custom repository.  and then modify the original code to query both places and merge the content.  You'll probably want to add an indicator that would show what repository it is from.

**Comment by Luraktinus** [comment]
any news there?

**Comment by hilleer** [comment]
@Luraktinus I haven't read any news regarding this, unfortunately.

**Comment by llgcode** [comment]
Also interested to have this feature.

**Comment by Luraktinus** [comment]
yeah, understandable

**Comment by gvanbeck** [comment]
+1

**Comment by bjoernbusch** [comment]
This would be really nice.

**Comment by notlaforge** [comment]
Is it possible to publish extensions on the marketplace and have them not appear in search?

I'm thinking for cases where the extension is not really secret, just not generally useful outside a company.

**Comment by sijakret** [comment]
+1
i believe this is an important feature for all sorts of custom toolings

**Comment by hilleer** [comment]
@Luraktinus so if you have a private extension that many employees are using, flow could be as following;

1. Distribute it to employees.
2. Employees all have to install it manually.

This, to me at least, includes a lot of time wasted, compared to just distributing it to something like the marketplace.

**Comment by Luraktinus** [comment]
@hilleer

I mean that the community could do a plug-in to replace the functionality of the gallery.

**Comment by levonmamikonyan** [comment]
As a stop gap we've setup an in-house webserver and hosted all extensions there. 
It also exposed an API endpoint to return latest number version for a given extension.
Then for every extension we had a timer to ping service every minute, check the version and compare with the one installed. 
So if a new version was detected it would prompt to upgrade and then from the it was just a bunch of chained promises (show a prompt, on OK download an extension, show a prompt to restart, restart).
Then the last step was to automate a build pipeline, so every extension release was properly deployed to webserver (really package the extension and copy to appropriate folder).
The only drawback is that the extensions must be installed manually for the first time, from there on it's all automated.

**Comment by jan-dolejsi** [comment]
Yeah. I have implement one of those custom plugin galleries before for
another extensible product I worked on. The plugins were managed in a
SharePoint document library. It dealt with the versioning, exposed a
RESTful interface, authentication, possibly user authorization. It worked.

But more and more enterprises move to O365. where it is suddenly much more
difficult for a desktop application to interact with. So if you tried to
implement this custom extension gallery based on any cloud hosted
repository with 2 factor authentication, you would have to take the user
through the authentication,
store the access token and refresh again when it expires...

The bottom line is that the biggest challenge is not to get the extension
to the right users, but to keep the extension up to date with subsequent
updates. What does not help is that the .vsix file does not _just install_
when you double click on it, but rather says it is invalid for any of the
_Visual Studio_ installation you have.
I realize I am not bringing any constructive suggestions here. Sorry.

Jan

On Mar 21, 2018 19:22, "levonmamikonyan" <notifications@github.com> wrote:

As a stop gap we've setup an in-house webserver and hosted all extensions
there.
It also exposed an API endpoint to return latest number version for a given
extension.
Then for every extension we had a timer to ping service every minute, check
the version and compare with the one installed.
So if a new version was detected it would prompt to upgrade and then from
the it was just a bunch of chained promises (show a prompt, on OK download
an extension, show a prompt to restart, restart).
Then the last step was to automate a build pipeline, so every extension
release was properly deployed to webserver (really package the extension
and copy to appropriate folder).
The only drawback is that the extensions must be installed manually for the
first time, from there on it's all automated.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/Microsoft/vscode/issues/21839#issuecomment-375047098>,
or mute the thread

**Comment by darkvertex** [comment]
Also interested in this. Would be super nice to have.

**Comment by itzik-h** [comment]
@seanmcbreen would you consider external help for that one?

**Comment by roysudi** [comment]
+1

**Comment by jkasun** [comment]
+1

**Comment by Ubeek** [comment]
+1

**Comment by oisteinwind** [comment]
This is a feature my team would very much like to use - it would be very much appreciated.

**Comment by acantrell** [comment]
+1

**Comment by roicpp** [comment]
+1

**Comment by Yinkai15** [comment]
+1

**Comment by swimmernewman** [comment]
+1

**Comment by ikbelkirasan** [comment]
+1

**Comment by hilleer** [comment]
Hi everyone,
As it seem many haven't realised, you can't just react to the original post with a thumb up. This will make it much easier to get an overview of how big a request this is, while you will not send notification to everyone following this post :)

**Comment by llgcode** [comment]
Hi,
Suggestion, I think most of people here don't need a complete marketplace. I think what we need first is a command that install an extension from a trusted url.

**Comment by naivefun** [comment]
Besides customized extensions, the private marketplace should be able to select subset of public extensions into it. That way we have a complete required extensions in one place. Of cos the public extensions are just a proxied uri.

**Comment by thalition** [comment]
Hello,
I am trying to get my VSCode to only being able to download extensions we in my team see fit our developers have. I have tried with changing some code in .\VSCode\resources\app\product.json where I deleted the link in "serviceUrl": "". This made it possible to deny our developers to just download extensions and meanwhile use the "install from VSIX" option.
I tried with changing it to links to our artifactory but it was not as simple as that, can anyone help wtih this? or will it first be able to be changed when this feature is implemented? :)

**Comment by hilleer** [comment]
@thalition weird to limit your developers to **your** needs and not theirs :) However, it doesn't sound related to the feature-request of this issue but as an entirely different scenario and feature.

**Comment by thalition** [comment]
@hiller Well its to start off with, the developers can come to us and tell us if they need other extensions, but its more so we can keep track of Who have what extension if something was to go wrong. We do this for all software :-) so its not to limit our developers just to know Who got what and why :-)
And as I understand this its a feature to make it possible to host a private internal marketplace, which i what i need, but I could be wrong :-)

**Comment by feldoh** [comment]
I agree with @thalition I think it's sort of the same thing. At the moment VSCode is bound to the main gallery. But we are air-gapped and thus a proxy is insufficient. So I'd also like to be able to link to our artifactory using something like the proposed additional extension gallery. We've got it set up for e.g. Maven, NPM, chocolatey, *nix packages etc so that your sources list (i.e. internal gallery) is artifactory and this has a mirror of the external repos as well as allowing us to have our own custom apps. As it stands I need to make a ticket for every extension I want or every update. If we could re-point the gallery to a mirror then we could install all the extensions available on the gallery as well as any internal ones.

**Comment by renanmt** [comment]
This is a much needed feature for my team. Can't wait for dev updates in this thread.

**Comment by hevayo** [comment]
We also have a similar requirement where we maintain a public plugin and need a developer channel to publish development version for a wider audience to test. So appreciate if you can add this feature.

**Comment by ajmas** [comment]
This would be useful in our form for the same reasons. Maybe a good starting place would to allow the market place URL to be overridable in the settings file? 

The next step would be to document the REST API. 

At this point either a community effort or a a company like JFrog (thinking of their Artifactory product) could provide support? — I have no vested interests in JFrog, just their solution is what we use in the firm I work for.

**Comment by haneefdm** [comment]
+1

**Comment by edopalmieri** [comment]
We would also really benefit from this. +1

**Comment by ajmas** [comment]
If you want to say you need this, then please add a thumbs up on the description of this ticket, rather adding a comment with a +1. 

Your vote in the description of the ticket adds more value. - Thanks

**Comment by rusnasonov** [comment]
It seems like [Autoupdating Google Chrome Extension](https://developer.chrome.com/apps/autoupdate)

1. Add `update_url` to package.json. It's a link to `update manifest`
2. Add `update manifest` support. Update manifest contains version and link to source file (.vsix)

It's a very simple and flexible workflow. For example, you can host own extension on Github easily  using Github Releases.

**Comment by gholt-aws** [comment]
+3 We need this for VS Code, Visual Studio and Azure DevOps.

**Comment by danielnmsft** [comment]
for me, the biggest advantage over vsix of being able to distribute via a private marketplace is auto-update functionality.  This is something I'm working on adding to my extension itself to get around the current limitation, but it would be great to take advantage of the marketplace so private extension updates can be automatically applied without any extra work.

**Comment by acohen716** [comment]
Any update on this?

**Comment by haydennyyy** [comment]
@acohen716 While MS could throw us a big wowie and make the marketplace open source, I doubt they will sadly. So... I doubt there's ever gonna be another official update for a _long_ time.

**Comment by chrisdias** [comment]
hi - its unlikely that we would open source the marketplace, but we are investigating the ability to have "unlisted" extensions. An unlisted extension would be in the marketplace, have a URL, but wouldn't be searchable. Once installed (manually), it would then get automatically updated as normal.

**Comment by nirnir93** [comment]
Would love to see support for non-official extension repository

**Comment by orelkan** [comment]
@chrisdias, while this solution solves the use case of a company developing private extensions, it doesn't solve the problem for companies working in closed networks.

We would appreciate it if we could have our own repository in the network

**Comment by syncush** [comment]
VSCode market should be open source and be able to run in closed networks.

**Comment by mdfederici** [comment]
> As a stop gap we've setup an in-house webserver and hosted all extensions there.
> It also exposed an API endpoint to return latest number version for a given extension.
> Then for every extension we had a timer to ping service every minute, check the version and compare with the one installed.
> So if a new version was detected it would prompt to upgrade and then from the it was just a bunch of chained promises (show a prompt, on OK download an extension, show a prompt to restart, restart).
> Then the last step was to automate a build pipeline, so every extension release was properly deployed to webserver (really package the extension and copy to appropriate folder).
> The only drawback is that the extensions must be installed manually for the first time, from there on it's all automated.

I'll probably need to end up doing this as well it seems like. I have private IP in the plugin which I cannot put in a public marketplace even if the plugin isn't searchable and I need the plugin to auto update at a weekly release cadence. @levonmamikonyan do you still have this solution and do you have any code you are willing/able to let me look at?

**Comment by bhunsicker** [comment]
@levonmamikonyan I have the same need as @mdfederici .  Are you willing to share some of that code you already have written?  hate reinventing the wheel :)

**Comment by levonmamikonyan** [comment]
@mdfederici @bhunsicker I will be glad to share the solution (there isn't much to brag about :)). Just have to get an all clear before we can open source it. Will let you know how it goes.

**Comment by fzs** [comment]
I'm also looking for a solution for an air-gapped network. We have set up local mirrors of Eclipse P2 repositories and for IntelliJ IDEA plugins, so that these extensions can be provided and updated at a central place and developers can install them into the IDE via the normal mechanism and also get updates.
We would need the same for VSCode, which has a high demand amongst developers. Even if it is just, like llgcode said, a local URL from that extensions can be installed.

**Comment by jamesharris-garmin** [comment]
It seems like the simplest solution would be to open source the API to the extensions library (publish official endpoint documentation) Then allow individuals to implement the API via tools like sonatype-nexus. This would allow us to configure and internally mirror/proxy extensions/publish extensions to an internal server.

**Comment by nirnir93** [comment]
+1 Would very much like to see this (I heard ProGet already hosts VSIXs... Is that related?)

**Comment by dmpe** [comment]
Exactly, simple plugin for sonatype nexus would be enough for us as well. This currently prohibits our company from rolling out VSCode "officially"

**Comment by yuxuan** [comment]
Base on this issue https://github.com/microsoft/vscode/issues/81173 answer, I'm not sure private marketplace would be allowed unless there is non-extensions from MS marketplace. 

My question is if we host a private extension server to serve internal extensions, and proxy other extensions to the official marketplace, does that violate MS terms.  

There may have same situations in this thread. So I post question here.

**Comment by rfresow** [comment]
We are air-gapped too and would like a solution whereby we can use Sonatype Nexus.

**Comment by cniweb** [comment]
@rfresow we too!

**Comment by vannaka** [comment]
A possible solution for yall. @rfresow, this extension manager **does** let you use Sonatype Nexus! https://github.com/joelspadin-garmin/vscode-private-extension-manager. I believe this extension will be in the official vscode marketplace soon.

**Comment by ttutko** [comment]
Also in an air-gapped situation and we have used https://github.com/LOLINTERNETZ/vscodeoffline.

So far it has worked well for us in a completely offline environment. The only gotcha that we have run into is that some extensions attempt to reach out and download additional things from github or elsewhere when first run which does not work if you are offline, obviously. For those cases (python plugin is an example) you have to have an online machine that generally matches the os/architecture of your air-gapped machines and trigger the download there then move the whole folder over from your extensions directory. In other cases such as the C++ extension, the developers detect the failed download and give you a message in the log output on how to download an offline package of all architectures. Would love to see more extension developers do something like that when their extension requires reaching out to download something.

**Comment by joelspadin-garmin** [comment]
The extension @vannaka linked above is now in the official marketplace at https://marketplace.visualstudio.com/items?itemName=garmin.private-extension-manager

**Comment by prashantvc** [comment]
Hey Everyone!

I am Prashant, PM on the VS Marketplace team. The team that serves extensions to VS Code, we are in the process of prioritising the VS Marketplace backlog, I would love to get fresh perspective on this issue (both enterprises and individual needs)

Please email me at prchol@microsoft.com, we can set up a quick call

**Comment by RamyElkest** [comment]
@prashantvc any insight on the approach being taken.

**Comment by prashantvc** [comment]
@RamyElkest We are evaluating the multiple scenarios. We do not have anything to update at the moment.

**Comment by mdschweda** [comment]
What about the announcement of Theia? This seems like an opportune moment to prevent future segmentation.

**Comment by mdschweda** [comment]
https://github.com/eclipse/openvsx

**Comment by joelspadin-garmin** [comment]
@prashantvc I'm not sure if I have time for a call anytime soon, but to answer your questions:

1. [I wrote an extension](https://marketplace.visualstudio.com/items?itemName=garmin.private-extension-manager) to manage other extensions from a private server.
2. The extensions are on a private NPM registry. We publish extensions as NPM packages to it.
3. Yes.
4. See answer 2 for publishing. For distributing, users either add the URL of the private server to a setting, or we add a file similar to `extensions.json` to our repos which has that URL and any recommended extensions. The Private Extension Manager lists the available extensions and handles installation and updates from there.

The goals of this solution are:

* Keep extensions which might contain proprietary info on a server we control.
* Make it easy to find and distribute extensions, especially extensions relevant to specific projects.
* Handle extension updates.
* Make it easy to find whatever does all of the above (hence making the extension manager open source, so we can recommend it in an `extensions.json`).

**Comment by jamesharris-garmin** [comment]
A further comment I would make regarding this: We tend to have a policy of wanting to do transparent mirroring proxies for mission critical resource, that way if something happens that takes down large sections of the internet we are insulated with our own mirrors. We achieve this using tools like Sonatype's [Nexus](https://www.sonatype.com/product-nexus-repository) or Jfrog's [Artifactory](https://jfrog.com/artifactory/)  so the ability to setup a similar capability would be useful to us.

**Comment by axi92** [comment]
We are using this https://github.com/joelspadin-garmin/vscode-private-extension-manager and a npm repo in nexus. Works very well!

**Comment by sr229** [comment]
Our friends from @eclipse Made OpenVSX. Give it a shot.

**Comment by Surendrajat** [comment]
Any updates on this one? I'm also looking for the same functionality.

**Comment by ludakhris** [comment]
Any official updates?

**Comment by yuzhidi** [comment]
> https://github.com/eclipse/openvsx

https://github.com/eclipse/openvsx/issues/116
So far, seems Open VSX is not really an alternative to the Visual Studio Marketplace, but rather meant for developers that are using gitpod to develop extensions.

**Comment by yuzhidi** [comment]
> Our friends from @eclipse Made OpenVSX. Give it a shot.

Do your friends deployed OpenVSX locally?

**Comment by yuzhidi** [comment]
> We are using this https://github.com/joelspadin-garmin/vscode-private-extension-manager and a npm repo in nexus. Works very well!

nexus should be paid?

**Comment by axi92** [comment]
> > We are using this https://github.com/joelspadin-garmin/vscode-private-extension-manager and a npm repo in nexus. Works very well!
> 
> nexus should be paid?
Nexus3 OSS is free to use =)

**Comment by svenefftinge** [comment]
> Do your friends deployed OpenVSX locally?

You can deploy it locally, yes.
https://github.com/eclipse/openvsx

**Comment by GitMensch** [comment]
> You can deploy it locally, yes.
https://github.com/eclipse/openvsx

... and then override marketplace in product.json for your users to point to your locally deployed OpenVSX installation. That way you check which extensions (and versions) you want to have, push those to your local open-vsx and let the users work with that.

For the initial user-setup I'd suggest to create an "empty" vscode project (or code-workspace) with `extensions.recommendations` setup to include all extensions the users should have (or create separate ones for the separate teams, for example one java/.vscode one python/.vscode and one common/.vscode; then depending on the team the users has to only open the correct workspace with the recommended extension setting, then clock on "yes install all recommended extensions" and is setup.
When you later push new versions to open-vsx they will be automatically updated.

**Comment by carmenbreezy** [comment]
Any updates on this?

**Comment by GitMensch** [comment]
If there's interest I'd create a PR to override the marketplace endpoints with environment variables (if they aren't set the product.json wins, I guess this may be acceptable).

**Comment by feldoh** [comment]
@GitMensch that would be an excellent tradeoff

**Comment by YonathanB** [comment]
4 years and still waiting for that feature!!

**Comment by kjsingh** [comment]
I am the 770th customer waiting for this feature!

**Comment by GitMensch** [comment]
OK, we have a PR to fix this >4 years issue now, _only_ needing someone to review and pull that.
If given some hints about where to add this I'll create a follow-up PR to vscode-docs...

This would really help to use vscode in environments that are disconnected from the internet and/or not allowed to install any extension they like, which is both not uncommon in corporate environments.

**Comment by prashantvc** [comment]
> OK, we have a PR to fix this >4 years issue now, _only_ needing someone to review and pull that.

Adding Sandeep's comment for visibility https://github.com/microsoft/vscode/pull/129764#issuecomment-889155248

**Comment by GitMensch** [comment]
@isidorn The upstream issue was closed one hour ago, does this mean a private marketplace/gallery is available now? Where exactly are the docs?

**Comment by GitMensch** [comment]
... I'm aware of the conversion which was done without any note about the reasons at all.

So there is an upstream discussion now, not even an "issue". Where does this lead this to, concerning this specific feature request with more than 800 user votes?

**Comment by AndersHogqvist** [comment]
bump!

**Comment by pthexton** [comment]
> @GitMensch
> 
> Converting the issue to discussion in no way undermines the importance of this request.
> 

There are 6 participants in that discussion, meanwhile this request has been starred over 127 thousand times. Out of sight out of mind? 

People (presumably within organisations, which is my use case) want this feature. VSCode is great for the ability to write your own productivity extensions, the requirement to publish to the public marketplace if you want to easily share with colleagues / simplify fresh environment setup simply does not make sense where those extensions are literally useless to anybody outside of a given development team / environment.

**Comment by StefanMaron** [comment]
> There are 6 participants in that discussion, meanwhile this request has been starred over 127 thousand times. Out of sight out of mind?

 I think the stars belong to the whole repository ;) 

But the rest you said, I support 💯

**Comment by pthexton** [comment]
> I think the stars belong to the whole repository ;)
> 


That's a great point, my bad! I've edited accordingly.

**Comment by GitMensch** [comment]
.. and there are still over 850 implementation votes here - because the people don't want a private marketplace support of the ms market (others exist and do work fine, so there's actual not much need for that "related" issue in the marketplace repo), but a simple way to connect to them... and I already forgot about this - I've created a PR to actually fix that (with open points like the docs) at https://github.com/microsoft/vscode/pull/129764 ...

**Comment by pthexton** [comment]
> .. and there are still over 850 implementation votes here - because the people don't want a private marketplace support of the ms market (others exist and do work fine, so there's actual not much need for that "related" issue in the marketplace repo), but a simple way to connect to them... and I already forgot about this - I've created a PR to actually fix that (with open points like the docs) at #129764 ...

Yeah. I literally just want to be able to publish a private vsix to an internal webserver and have a way of telling vscode "please also look in here", that way any individual projects can use the extensions.json recommendations as normal.

**Comment by GitMensch** [comment]
Work around:

* setup the server, for example https://github.com/eclipse/openvsx locally
* adjust package.json of every vscode installation to point to this local server
* publish your "local/team" extensions there
* publish every other extension there that you want to use

... but your scenario actually brings up another point, It seems you likely want to be able to reach multiple servers (which would remove the need to publish "every other allowed" extensions to your local server) - that's something reasonable but should be tracked under a new issue (if it doesn't exist) "Support for multiple marketplaces/extension galleries".

**Comment by pthexton** [comment]
@GitMensch I think my scenario is covered by this part of the OP

> and have an ability to specify alternative extension gallery paths

**Comment by lextm** [comment]
Even if a private marketplace/gallery is made available (like openvsx), enterprises with strict network management will still find gaps running it nicely with VSCode and the extensions (especially when many extensions try to download from external).

My company LeXtudio Inc. is aiming to build an end-to-end solution to enable enterprises to set up private VSCode extension marketplace/gallery, and harden private extensions (so they don't download anything from external).

If you are interested in such a product, we are in private beta phase right now. You can write to support@lextudio.com to learn how to request a demo or join the beta program.

**Comment by codecycleteam** [comment]
@isidorn work in progress or not bro? If yes then any ETA?

**Comment by GreyElec** [comment]
Vote for this feature.

**Comment by uberDoward** [comment]
Still waiting…

**Comment by rlgloden** [comment]
+1
This is really needed of air gapped networks.

**Comment by plffy** [comment]
+1

**Comment by TheMasterFX** [comment]
In our company we also need this feature since our working PCs do not have any connection to the internet (only company intranet). It is very frustrating and not to handle for our IT-Department to supply the needed plugins to the client computers.

**Comment by xinghun** [comment]
Vote for this feature.

**Comment by z-juln** [comment]
This is my solution: <https://juejin.cn/post/7141662937420136479/>

**Comment by ghuntley** [comment]
Hello folks 👋 from Coder. Today we open-sourced "code-marketplace" which is an alternative to OpenVSX that provides an easier way to self-host your own VSCode marketplace. Maybe it'll help with all of your with your needs re: AirGapping (it's why we authored it in the first place!)

It's is a self-contained go binary that does not have a frontend or any mechanisms for extension authors to add or update extensions in the marketplace. It simply reads extensions from file storage and provides an API for VSCode compatible editors to consume.

Details at https://coder.com/blog/running-a-private-vs-code-extension-marketplace and source at https://github.com/coder/code-marketplace

**Comment by GitMensch** [comment]
@sandy081 closed the related #129764 because

> this is in our and Marketplace team's road map

which was back in 2021. As there is "still nothing concrete" soon 2 years later - maybe someone could re-review that PR?

I do agree with @timheuer that (additional) an option to specify a path to vsix files would be very useful, while this would mean that vscode would have to unpack at least part of it in a cached directory (README, possibly CHANGELOG, product.json, ...) to allow the extension view to actually search and view those.

**Comment by isidorn** [comment]
@timheuer thanks for commenting. Can you open a new issue and ping me please? This issue has a lot of requests and I would like to keep these separate since it would be easier to manage. Thank you 🙏

**Comment by Rutix** [comment]
@isidorn Is there any update though on the requests in this issue? Like @GitMensch says @sandy081 closed previous attempts ages ago and pinged @prashantvc but we have not received any update since.

**Comment by prashantvc** [comment]
@Rutix I no longer works with VS Marketplace team. 

But, I have something in works if you want to give it a try 😊

**Comment by ajmas** [comment]
For someone who isn’t actively working on the project, can anyone who knows indicate whether the issue is technical or policy?

**Comment by prashantvc** [comment]
Just published a preview of private repository here → https://github.com/prashantvc/extensions/releases/tag/v0.0.1
Give it a try, file an issue or send a PR :)

**Comment by timheuer** [comment]
@prashantvc cool! This only solves a vsix host right and doesn’t have a means to attach to a marketplace client in vscode.

**Comment by prashantvc** [comment]
@timheuer 

> @prashantvc cool! This only solves a vsix host right and doesn’t have a means to attach to a marketplace client in vscode.

Yes, that is right. VS Code does not allow anyone to extend marketplace client (ie. extension view) in VS Code, and not to mention marketplace APIs are hot mess to clone.

The workaround is to build a new view (client), like many other solutions in this issue. 
You can switch to `vscode-extension` branch in [extensions repo](https://github.com/prashantvc/extensions) to try the one I have in works.

Here's a quick preview: 
<img width="600" alt="image" src="https://user-images.githubusercontent.com/328076/235718710-44fe3ade-920f-4a6a-aba0-13e2f50567bf.png">

**Comment by GitMensch** [comment]
> VS Code does not allow anyone to extend marketplace client (ie. extension view) in VS Code

... which would not necessary if one could override the target endpoints - see #129764 from 2021 for a _working solution_ (which you can use if you build vscode from scratch); or if vscode would allow to specifiy multiple/additional endpoints.

> and not to mention marketplace APIs are hot mess to clone.

... there are several projects mentioned here that exactly do this, including osvx-registry (the software, not the public instance run by Eclipse).

> For someone who isn’t actively working on the project, can anyone who knows indicate whether the issue is technical or policy?

That's more @isidorn to answer but @sandy081 closed the mentioned PR and similar issues because "the marketplace team has come up with a plan" two years ago, which they seem to have dropped. Maybe it is time to re-review the PR (if it is reopened I'd update it, of course)?

**Comment by jeanp413** [comment]
@timheuer @prashantvc is/will https://github.com/prashantvc/extensions officially mantained by microsoft or it's just another free alternative like the ones already mentioned in this issue?

**Comment by prashantvc** [comment]
@jeanp413 it’s one of my side projects that I am working. The project is not connected to Microsoft in anyway

**Comment by msweeney1999** [comment]
I too need this feature.

**Comment by Dongyang0810** [comment]
+1

**Comment by maxenium** [comment]
I strongly support the idea of implementing a private extensions store for Visual Studio Code. As a representative of many companies and organizations, we often find ourselves in a situation where we need to develop and use in-house extensions and tools, but sharing them on the public marketplace is simply not an option due to sensitive data and intellectual property concerns.

The current landscape has led to workarounds such as relying on third-party plugins or tools to create private extensions stores e.g. [Private Extension Manager](https://marketplace.visualstudio.com/items?itemName=garmin.private-extension-manager). However, these solutions often come with their own set of problems. Many of these third-party plugins are no longer actively maintained, and they tend to break with newer versions of VS Code and its dependencies. This makes managing private extensions within a corporate environment impossible without going the manual route of downloading vsix files.

One significant issue we've encountered is the configuration of proxy settings. Outdated extensions fail to recognize and adapt to the latest proxy configurations, which can lead to connectivity issues and security vulnerabilities. For instance, newer versions of VS Code do not work with the "Private Extensions Manager" and the 'no_proxy' settings are not being recognized, causing problems when working behind a proxy. 

A dedicated private extensions store within VS Code would address these issues and provide a secure and reliable way for companies to make their proprietary extensions available to their employees. It would enhance the development workflow, streamline the distribution of internal tools, and ultimately improve productivity and collaboration within the organization.

I believe this feature would be a valuable addition to Visual Studio Code, benefiting both individual developers and enterprises alike.

I am wondering why this feature request gets so little attention from the development team. 😢

**Comment by Zane-XY** [comment]
Need this feature! Super important for enterprise users. May just reference what IntelliJ platform does.

**Comment by taturou** [comment]
I would appreciate this feature.
I am creating private tools, and distributing updates is very tedious.

**Comment by omerd-cyera** [comment]
+1

**Comment by gjsjohnmurray** [comment]
Se also https://github.com/microsoft/vscode/issues/179919

**Comment by Jonas56** [comment]
This feature would be very helpful for teams with closed networks. 

Currently, our method involves utilizing GitHub releases to publish the latest versions, followed by manual notifications to users upon each update—an approach that is really annoying. I propose the incorporation of a feature allowing the customization or override of the default [marketplace](https://github.com/microsoft/vscode/blob/578ad35313f41c4aeff777ab25bd3265c645ab34/build/lib/extensions.ts#L228) link within VSCode. Then, introduce the capability to integrate additional links to alternative sources, such as Artifactory, NPM, GitHub releases, etc.

**Comment by zentby** [comment]
Try integrate this [Auto Updater](https://marketplace.visualstudio.com/items?itemName=yangzhao.auto-updater&ssr=false#overview) plugin into your extension.

With this extension, your extension can be hosted in your private website(e.g. behind a VPN).

**Comment by seaniyer** [comment]
@JacquelineWiddis

**Comment by Rafik-Belkadi-malou** [comment]
Need this af

**Comment by GitMensch** [comment]
> Need this af

posting here obviously don't help... either you build vscode yourself, applying the related PR to run a private instance for example of openvsx (that PR was closed because "the marketplace team has come up with a plan" nearly 3 years ago - we all see how well and timely that worked out...), or use vscodium, which did accept a similar patch, or use something like https://marketplace.visualstudio.com/items?itemName=zokugun.vsix-manager which allows to install extensions from any workplace or locally (but needs manual setup of the extensions).

**Comment by omerd-cyera** [comment]
I ended up creating releases in github that have the vsix file, and created a component inside the extension that checks for updates periodically, and updates itself.

Not perfect, but its simpler than you think to implement, and it doesn't require extension consumers anything apart for installing the extension from vsix once.

**Comment by fone-almosca** [comment]
@omerd-cyera That's exactly what i'm trying to do for the past 2 days, but when I arrive to the point of uploading the .vsix file into release asset, i get the error: 

> "Resource not accessible by integration" 

I'm pretty sure the token as the correct rights, as the upload works for every file except .vsix. 
So I'm wondering, if you had the same issue, how did you solve it ?

**Comment by omerd-cyera** [comment]
@fone-almosca I think vsix is not supported in gh. Just zip or tar it.

**Comment by gaby** [comment]
7 years later and this issue still open? @isidorn Any updates on this?

**Comment by Lisoph** [comment]
> I ended up creating releases in github that have the vsix file, and created a component inside the extension that checks for updates periodically, and updates itself.

How did you accomplish the automatic updating? Do you just invoke the `workbench.extensions.action.installVSIX` command?

**Comment by omerd-cyera** [comment]
I have other tools installed on all relevant machines that are responsible for first install and updates. In my case it was easier to just trigger them via the cli.
I think that in most cases `workbench.extensions.action.installVSIX ` is the way to go.

**Comment by jameswiseman76rlg** [comment]
In light of the following article, wouldn't you say that the need for this feature becomes more pressing?

https://www.bleepingcomputer.com/news/security/malicious-vscode-extensions-with-millions-of-installs-discovered/amp/

_"Microsoft's lack of stringent controls and code reviewing mechanisms on the VSCode Marketplace allows threat actors to perform rampant abuse of the platform, with it getting worse as the platform is increasingly used."_

I work for an org that hosts private NPM, NUGET and Visual Studio extension gallery where we curate the packages we host. It feels like a natural progression to allow this for VS code, especially with with integration through ADO Artefacts.

**Comment by jameswiseman76rlg** [comment]
https://github.com/microsoft/vscode/issues/84756#issuecomment-2157964224

![image](https://github.com/microsoft/vscode/assets/153926637/5cd75889-9d46-429a-bf7c-1fb38f54862b)

**Comment by swythan** [comment]
> As already noted: you can use and curate your own extension marketplace by setting up a "local" entry point target, for example an [OpenVSX](https://github.com/eclipse/openvsx) instance and adjusting package.json to point to that. 

I guess this was in response to the [preceding comment](https://github.com/microsoft/vscode/issues/21839#issuecomment-2157691569). Just to be clear though, this wouldn't solve the problem in the OP (which I share): to have an internal marketplace _in addition to_ the main public marketplace.

**Comment by danielboucek** [comment]
For those of you who use Gitlab. I recently created an extension that uses Gitlab's Package Registry API. Allowing you to browse, install, and auto-update private extensions.
You can check it out if you like. It works surprisingly well. 
[Private Extension Manager - GitLab](https://marketplace.visualstudio.com/items?itemName=DanielBoucek.private-extensions-gitlab)

**Comment by sameemqureshi** [comment]
We are setting up our private Extension on JFrog Artifactory , For the First time the Users would be getting the Extension from the Artifactory itslef.
there are going to be updates for this extension , we want to implement a seamless update flow for the users , where there extension would be pinging too check for any updates and If found , notify and update the extension,
How can this Update flow be seamless , where user just needs o click the update , and latest version is taken?
how can it be automated??

https://github.com/microsoft/vscode/issues/21839#issuecomment-375047098

**Comment by seaniyer** [comment]
> Hi, before I open your URL, I can see that you are not a "member" of this project. How can we verify you?

Sorry to hear that. This may be a localized issue as few others were already able to take the survey. @deadmeu, one suggestion is to try the link from an InPrivate browser window.

**Comment by omerd-cyera** [comment]
I fully agree with @deadmeu, I'm also not seeing a "member" tag. Also, opening in a private browser window isn't really reassuring

**Comment by isidorn** [comment]
@seaniyer works on our VS Marketplace. I am a PM working on VS Code.
You can trust the link he posted.

**Comment by jvilk-stripe** [comment]
I filled out the survey. Feel free to reach out if you want to chat more about our use cases at Stripe.

**Comment by isidorn** [comment]
Thank you very much for your replies in the survey.
In the survey, I noticed that some of you would like an allow list of extensions from the public marketplace. Please be aware that we are launching this feature next week in VS Code Insiders. And more details can be found here https://github.com/microsoft/vscode/issues/84756

Feedback is very much welcome!

**Comment by kineticsquid** [comment]
We think the idea of a private marketplace is a great idea. In fact, we have one of these at the Eclipse Foundation Open VSX Registry at [open-vsx.org](http://open-vsx.org/). It's an open, transparent, vendor neutral registry for VS Code extensions comprised of three Eclipse open source projects and governed by the Eclipse [Open VSX Working Group](https://projects.eclipse.org/working-group/open-vsx).
- https://github.com/EclipseFdn/open-vsx.org
- https://github.com/eclipse/openvsx
- https://github.com/open-vsx/publish-extensions

Easily deployable in house in a tethered or untethered configuration, including in an air-gapped environment. You can try it out quickly locally or with a [couple of clicks at Gitpod](https://gitpod.io/#https://github.com/eclipse/openvsx). We also have a [PR](https://github.com/eclipse/openvsx/pull/1011) coming through to provide the same for Red Hat OpenShift.

**Comment by hushoca** [comment]
Any updates on this?

**Comment by Lordirockz-ship-it** [comment]
1. > [`/****`](

[_r1j_-transcript.txt](https://github.com/user-attachments/files/19684180/_r1j_-transcript.txt)

)

**Comment by seaniyer** [comment]
Hi everyone, 

### **[Sign up](https://www.surveymonkey.com/r/SF7YMVZ?src=gh)** today for a chance to participate in the upcoming preview of Private Marketplace for VS Code! 

VS Code Marketplace is planning a preview program in May with about 50 customers. The preview release will be a container for self-hosting internal extensions in a file system directory or Azure Artifacts feed. Features will include upstreaming for automatic inclusion of public extensions, installation and auto-update of private extensions in VS Code. The preview will be available soon after the next release of VS Code. See the sign-up link for more details on capabilities included.

VS Code Marketplace team

**Comment by Mike-FreeAI** [comment]
> Hi everyone,
> 
> ### **[Sign up](https://www.surveymonkey.com/r/SF7YMVZ?src=gh)** today for a chance to participate in the upcoming preview of Private Marketplace for VS Code!
> VS Code Marketplace is planning a preview program in May with about 50 customers. The preview release will be a container for self-hosting internal extensions in a file system directory or Azure Artifacts feed. Features will include upstreaming for automatic inclusion of public extensions, installation and auto-update of private extensions in VS Code. The preview will be available soon after the next release of VS Code. See the sign-up link for more details on capabilities included.
> 
> VS Code Marketplace team

yay

**Comment by mschan01** [comment]
Is there any news from the feature preview?  My organization is very interested in trying out this feature.

**Comment by seaniyer** [comment]
@mschan01 
It is currently in a limited Preview. Please be sure to [sign up](https://www.surveymonkey.com/r/SF7YMVZ?src=gh) for the preview with your organization name. If you have recently done so, thank you and we'll follow-up. You can also reach out privatemktplace@microsoft.com for 1:1 help.

**Comment by Pargit7** [comment]
Hello,
Do you know if there's a release date, even a tentative one?

Thank you.

**Comment by seaniyer** [comment]
We are still planning the General Availability date.  I'll post it here as soon as available.

**Comment by FrsECM** [comment]
Any update on this thread ?
It would be very helpfull in enterprise environment when we **must** use secured mirrors with vulnerability scans.

Thanks !

**Comment by swythan** [comment]
As asked for in the OP, part of the request here is to have a private way to distribute extensions _in addition to_ what is on the public marketplace. OpenVSX doesn't really help much as you'd need a system for keeping it up to date with the public marketplace (or perhaps some subset of it).

Note this article about leaked secrets in VS Code extensions specifically calls out the problem with putting company-specific extensions in the public marketplace:

>An additional interesting lens on these leaked tokens involves the public distribution of company internal or vendor specific extensions. If you investigate the marketplace, you’ll notice extensions that have a low install count, but are specifically designed to support a single company’s engineers or customers. _**Internal extensions should not be distributed publicly, but often are for convenience.**_ In one case, we found a VSCode Marketplace PAT that would allow us to push targeted malware to the workforce of a $30 billion market cap Chinese megacorp. Vendor specific extensions are common, and allow for interesting targeting opportunities if compromised. For example, one at risk extension belonged to a Russian construction technology company. 

https://www.wiz.io/blog/supply-chain-risk-in-vscode-extension-marketplaces (emphasis mine)

**Comment by seaniyer** [comment]
Thanks @Foda 

We are excited to announce that the VS Code Private Marketplace is now generally available!

Enterprises can self-host and control extension distribution for their dev teams and dynamically include select/all public extensions . Secure, compliant, and seamless - right inside VS Code. 
Start here to explore more [blog post:](https://code.visualstudio.com/blogs/2025/11/18/PrivateMarketplace)

For help reach out to https://aka.ms/vspm/support

Happy coding!

**Comment by DisgruntledCrayon23** [comment]
@seaniyer @Foda,

Very exciting news! Thanks for the update!

If I'm understanding correctly, this feature only available with a licensed connection to GitHub Enterprise? Is that correct?

If so, that wouldn't be very helpful for air-gapped networks... Are there any plans to extend the functionality beyond its current state?

**Comment by GitMensch** [comment]
> If I'm understanding correctly, this feature only available with a licensed connection to GitHub Enterprise? Is that correct?
> 
> If so, that wouldn't be very helpful for air-gapped networks... Are there any plans to extend the functionality beyond its current state?

Do we need to create a sub issue for that (because this issue is now closed an the companies I work with don't have any GitHub at all, they just want to use a "local" marketplace entry point.
